docs/proposal/1-authentication-support/README.md

```diff
@@ -0,0 +1,149 @@
+
+## Goals
+
+1. Use username/password authentication to implement login feature(create an admin user by default). 
+2. Support extendable authentication protocol (e.g. ldap, oidc).
+ 
+## Non-Goals
+
+Implement a complex RBAC authentication system.
+
+## Use Cases
+
+Visiting [kstone-dashboard](https://github.com/tkestack/kstone-dashboard) home page, it redirects to the login page.
+
+### Basic Auth
+
+Users use their username and password to login.
+
+### External Auth
+
+Users configure their own authentication system (e.g. ldap, oidc), and use their custom authentication system to verify identity.
+
+## Proposal
+
+In order to implement login feature, we develop an authentication module in kstone-api. It supports not only basic auth(username/password authentication) by default but also other authentication protocols.
+
+## Design Details
+
+### Authentication Interface
+
+```go
+
+// LoginUser includes user login information
+type LoginUser struct {
+    Username   string
+    Password   string
+    User       *User
+    // external auth module (e.g. ldap, oidc)
+    AuthModule string
+    // Cfg includes the external auth config
+    Cfg        *AuthCfg
+}
+
+// Authentication is an authentication interface for kstone-api
+type Authentication interface {
+
+	// Authenticate does authentication based on given LoginUser instance
+	Authenticate(ctx context.Context, loginUser LoginUser) (*AuthenticateResponse, error)
+
+	// UserAdd adds a new user
+	UserAdd(r *AuthUserAddRequest) (*AuthUserAddResponse, error)
+
+	// UserDelete deletes a user
+	UserDelete(r *AuthUserDeleteRequest) (*AuthUserDeleteResponse, error)
+
+	// UserChangePassword changes a password of a user
+	UserChangePassword(r *AuthUserChangePasswordRequest) (*AuthUserChangePasswordResponse, error)
+
+	// UserGet gets the detailed information of a users
+	UserGet(r *AuthUserGetRequest) (*AuthUserGetResponse, error)
+	
+	// UserList gets a list of all users
+	UserList(r *AuthUserListRequest) (*AuthUserListResponse, error)
+
+	// CheckPassword checks a given pair of username and password is correct
+	CheckPassword(username, password string) (uint64, error)
+	
+	// BcryptCost gets strength of hashing bcrypted auth password
+	BcryptCost() int
+}
+```
+
+### Basic Auth
+
+- Auth Configuration
+
+By default, we create a `kstone-api-admin` secret in kstone namespace, which specifies admin user and password.
+
+```yaml
+apiVersion: v1
+data:
+  # bcrypt encryption
+  password: $2y$10$y7QLHNLpMvn7SJmrqNhYBe3YHnXuLPOYzemJBXU9Bownp4RNAlkQy
+  # base64 encoding
+  user: YWRtaW4=
+kind: Secret
+metadata:
+  name: kstone-api-admin
+  namespace: kstone
+type: Opaque
+```
+
+To customize admin username and password, modify the user and password field in `kstone-api-admin` secret.
+
+- Encryption && Decryption
+
+Use [`golang.org/x/crypto`](https://pkg.go.dev/golang.org/x/crypto) to encrypt the password:
+
+```go
+bcrypt.GenerateFromPassword([]byte(password), as.bcryptCost)
```

Comment:
does each request need to verify the password once? performance is very poor.
refer to the etcd authentication implementation.
https://github.com/etcd-io/etcd/tree/main/server/auth